src/librustc/hir/lowering.rs

```diff
@@ -437,8 +446,42 @@ impl<'a> LoweringContext<'a> {
     }
 
     fn lower_generics(&mut self, g: &Generics) -> hir::Generics {
+        // Collect `?Trait` bounds in where clause and move them to parameter definitions.
+        let mut add_bounds = NodeMap();
+        for pred in &g.where_clause.predicates {
+            if let WherePredicate::BoundPredicate(ref bound_pred) = *pred {
+                'next_bound: for bound in &bound_pred.bounds {
+                    if let TraitTyParamBound(_, TraitBoundModifier::Maybe) = *bound {
+                        let report_error = || {
+                            self.diagnostic().span_err(bound_pred.bounded_ty.span,
+                                "`?Trait` bounds are permitted only on fresh type parameters");
+                        };
+                        // Check if the where clause type is a plain type parameter.
+                        match bound_pred.bounded_ty.node {
+                            TyKind::Path(None, ref path)
+                                    if !path.global && path.segments.len() == 1 &&
+                                        bound_pred.bound_lifetimes.is_empty() => {
+                                let name = path.segments[0].identifier.name;
+                                for ty_param in &g.ty_params {
+                                    // Don't bother looking up the definition,
```

Comment:
Sure. Randomly implementing unhygienic lookup is a sure way to cause bugs in the future.